sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/properties/AadAuthenticationProperties.java

```diff
@@ -552,100 +568,115 @@ private void validateApplicationType() {
+     * "offline_access" : allows to request a refresh token.
+     * @param properties AuthorizationClientProperties
+     * @param scopes scopes for authorization_code clients.
+     */
+    private void addNecessaryScopesForAuhtorizationCodeClients(AuthorizationClientProperties properties,
+                                                               List<String> scopes) {
+        if (properties.getAuthorizationGrantType().equals(AUTHORIZATION_CODE)) {
+            String[] scopesNeeded = new String[] { "openid", "profile", "offline_access" };
+            for (String scope : scopesNeeded) {
+                if (!scopes.contains(scope)) {
+                    scopes.add(scope);
+                }
             }
         }
+    }
 
-        // Validate scopes.
+    private List<String> extractValidatedScopes(String registrationId, AuthorizationClientProperties properties) {
         List<String> scopes = properties.getScopes();
         if (scopes == null || scopes.isEmpty()) {
             throw new IllegalStateException(
                 "'spring.cloud.azure.active-directory.authorization-clients." + registrationId + ".scopes' must be "
                     + "configured");
         }
-        // Add necessary scopes for authorization_code clients.
-        // https://docs.microsoft.com/graph/permissions-reference#remarks-17
-        // https://docs.microsoft.com/azure/active-directory/develop/v2-permissions-and-consent#openid-connect-scopes
-        if (properties.getAuthorizationGrantType().getValue().equals(AUTHORIZATION_CODE.getValue())) {
-            if (!scopes.contains("openid")) {
-                scopes.add("openid"); // "openid" allows to request an ID token.
-            }
-            if (!scopes.contains("profile")) {
-                scopes.add("profile"); // "profile" allows returning additional claims in the ID token.
-            }
-            if (!scopes.contains("offline_access")) {
-                scopes.add("offline_access"); // "offline_access" allows to request a refresh token.
+        return scopes;
+    }
+
+    private void validateAuthorizationGrantType(String registrationId, AadAuthorizationGrantType grantType,
+                                                AuthorizationClientProperties properties) {
+
+        if (compatibleApplicationTypeAndGrantTypes.containsKey(applicationType)) {
+            if (compatibleApplicationTypeAndGrantTypes.get(applicationType).contains(grantType)) {
+                throw new IllegalStateException(String.format(UNMATCHING_OAUTH_GRANT_TYPE_FROMAT,
+                    applicationType.getValue(), registrationId, grantType));
             }
+            LOGGER.debug("'spring.cloud.azure.active-directory.authorization-clients.{}.authorization-grant-type'"
+                + " is valid.", registrationId);
+        }
+
+        if (AZURE_CLIENT_REGISTRATION_ID.equals(registrationId)
+            && AUTHORIZATION_CODE != properties.getAuthorizationGrantType()) {
```

Comment:
Doesn't this `properties.getAuthorizationGrantType()` equal to the given `grantType`?

Reply:
and?...

Reply:
then why should we pass in the properties, we can use the grantType directly

Reply:
yes